moshi-kotlin/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapterFactory.kt

```diff
@@ -141,51 +140,64 @@ internal class KotlinJsonAdapter<T>(
     for (binding in allBindings) {
       if (binding == null) continue // Skip constructor parameters that aren't properties.
 
-      writer.name(binding.jsonName)
+      writer.name(binding.name)
       binding.adapter.toJson(writer, binding.get(value))
     }
     writer.endObject()
   }
 
-  override fun toString() = "KotlinJsonAdapter(${constructor.returnType})"
+  override fun toString() = "KotlinJsonAdapter(${constructor.type.canonicalName})"
 
   data class Binding<K, P>(
+    val name: String,
     val jsonName: String,
     val adapter: JsonAdapter<P>,
-    val property: KProperty1<K, P>,
-    val parameter: KParameter?,
-    val propertyIndex: Int,
+    val property: KtProperty,
+    val propertyIndex: Int = property.parameter?.index ?: -1,
   ) {
-    fun get(value: K) = property.get(value)
+
+    fun get(value: K): Any? {
+      property.jvmGetter?.let { getter ->
+        return getter.invoke(value)
+      }
+      property.jvmField?.let {
+        return it.get(value)
+      }
+
+      error("Could not get JVM field or invoke JVM getter for property '$name'")
+    }
 
     fun set(result: K, value: P) {
       if (value !== ABSENT_VALUE) {
-        (property as KMutableProperty1<K, P>).set(result, value)
+        property.jvmSetter?.let { setter ->
+          setter.invoke(result, value)
+          return
```

Comment:
I don’t love `return` inside `let` because I cannot remember how much it returns